index/scorch/segment/zap/docvalues.go

```diff
@@ -240,34 +246,49 @@ func (di *docValueReader) getDocValueLocs(docNum uint64) (uint64, uint64) {
 
 // VisitDocumentFieldTerms is an implementation of the
 // DocumentFieldTermVisitable interface
-func (s *SegmentBase) VisitDocumentFieldTerms(localDocNum uint64, fields []string,
-	visitor index.DocumentFieldTermVisitor) error {
-	var dvIterClone *docValueReader
-	fieldIDPlus1 := uint16(0)
-	ok := true
+func (s *Segment) VisitDocumentFieldTerms(localDocNum uint64, fields []string,
+	visitor index.DocumentFieldTermVisitor, dvsIn segment.DocVisitState) (
+	segment.DocVisitState, error) {
+	dvs, ok := dvsIn.(*docVisitState)
+	if !ok || dvs == nil {
+		dvs = &docVisitState{
+			dvrs: make(map[uint16]*docValueReader, len(fields)),
+		}
+	} else {
+		if dvs.segment != s {
+			dvs.segment = s
+			dvs.dvrs = make(map[uint16]*docValueReader, len(fields))
+		}
+	}
+
+	// find the chunkNumber where the docValues are stored
+	docInChunk := localDocNum / uint64(s.chunkFactor)
+	var dvr *docValueReader
+	var fieldIDPlus1 uint16
 	for _, field := range fields {
 		if fieldIDPlus1, ok = s.fieldsMap[field]; !ok {
 			continue
 		}
-		// find the chunkNumber where the docValues are stored
-		docInChunk := localDocNum / uint64(s.chunkFactor)
-
-		if dvIter, exists := s.fieldDvReaders[fieldIDPlus1-1]; exists &&
+		fieldID := fieldIDPlus1 - 1
+		if dvIter, exists := s.fieldDvReaders[fieldID]; exists &&
 			dvIter != nil {
-			dvIterClone = dvIter.cloneInto(dvIterClone)
+			if dvr, ok = dvs.dvrs[fieldID]; !ok {
+				dvr = dvIter.cloneInto(dvr)
```

Comment:
one thought is to move the initialization of the dvs.dvrs to earlier, like up to line 260.5, where you know it needs to be initialized because dvrs is newly created.  Then, that would save an extra map lookup here at line 273 of s.fieldDvReaders[fieldID], where instead you'd only need to do a single lookup into the dvs.dvrs[fieldID] map.